### PR TITLE
[SPARK-16371][SQL] Do not push down filters incorrectly when inner name and outer name are the same in Parquet

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Currently, if there is a schema as below:

```
root
  |-- _1: struct (nullable = true)
  |    |-- _1: integer (nullable = true)
```

and if we execute the codes below:

``` scala
df.filter("_1 IS NOT NULL").count()
```

This pushes down a filter although this filter is being applied to `StructType`.(If my understanding is correct, Spark does not pushes down filters for those).

The reason is, `ParquetFilters.getFieldMap` produces results below:

```
(_1,StructType(StructField(_1,IntegerType,true)))
(_1,IntegerType)
```

and then it becomes a `Map`

```
(_1,IntegerType)
```

Now, because of `....lift(dataTypeOf(name)).map(_(name, value))`, this pushes down filters for `_1` which Parquet thinks is `IntegerType`. However, it is actually `StructType`.

So, Parquet filter2 produces incorrect results, for example, the codes below:

```
df.filter("_1 IS NOT NULL").count()
```

produces always 0.

This PR prevents this by not finding nested fields.
## How was this patch tested?

Unit test in `ParquetFilterSuite`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Hi, @rxin @liancheng, I hope this is not missed to 2.0..

**Comment by HyukjinKwon** [comment]
cc @viirya as well.

**Comment by liancheng** [comment]
LGTM pending Jenkins.

2.0.0 RC2 has already been cut. We may have this in 2.0.0 if there was another RC.

**Comment by HyukjinKwon** [comment]
Oh, @liancheng I just corrected some more. Please take another look..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 1]
```diff
@@ -188,7 +188,7 @@ private[sql] object ParquetFilters {
       fields.filter { f =>
         !f.metadata.contains(StructType.metadataKeyForOptionalField) ||
           !f.metadata.getBoolean(StructType.metadataKeyForOptionalField)
-      }.map(f => f.name -> f.dataType) ++ fields.flatMap { f => getFieldMap(f.dataType) }
+      }.map(f => f.name -> f.dataType)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: Could you please add some comment here?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: Sure!

**Comment by viirya** [comment]
The description seems incorrect. It should be a StringType, instead of IntegerType?

**Comment by HyukjinKwon** [comment]
yes it is, I just found I will correct them all. Thank you!

**Comment by viirya** [comment]
Another question is when the inner field is not the same name, we still can push down the filter, right?

**Comment by viirya** [comment]
If so, then this patch seems completely skip all such push down.

**Comment by HyukjinKwon** [comment]
Hm... I though Spark does not support filter-push down for nested fields.

**Comment by liancheng** [comment]
Yea, currently Spark SQL doesn't support column pruning and/or filter push-down for nested fields.

**Comment by viirya** [comment]
OK. LGTM then.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 2]
```diff
@@ -543,4 +543,18 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
       }
     }
   }
+
+  test("Do not push down filters incorrectly when inner name and outer name are the same") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: add the jira number here

**Comment by rxin** [comment]
I'm going to merge this and fix some comments myself with another pr. Merging in master/2.0.

**Comment by rxin** [comment]
FYI https://github.com/apache/spark/pull/14074/files

**Pull request closed without merging**